src/passes/J2CLItableMerging.cpp

```diff
@@ -0,0 +1,347 @@
+                                    .getStruct()
+                                    .fields[i - 1]
+                                    .type.getHeapType()));
+          }
+        }
+      }
+    };
+
+    Reindexer reindexer(*this);
+    reindexer.run(getPassRunner(), &wasm);
+    reindexer.runOnModuleCode(getPassRunner(), &wasm);
+  }
+
+  // Redirects all itable access by changing `struct.get` of the `itable` field
+  // to `struct.get` on the to `vtable` field.
+  void rerouteItableAccess(Module& wasm) {
+    struct Rerouter : public WalkerPass<PostWalker<Rerouter>> {
+      bool isFunctionParallel() override { return true; }
+
+      J2CLItableMerging& parent;
+
+      Rerouter(J2CLItableMerging& parent) : parent(parent) {}
+
+      std::unique_ptr<Pass> create() override {
+        return std::make_unique<Rerouter>(parent);
+      }
+
+      void visitStructGet(StructGet* curr) {
+        if (curr->ref->type == Type::unreachable) {
+          return;
+        }
+
+        if (!curr->type.isStruct() ||
+            !parent.structInfoByITableType.count(curr->type.getHeapType())) {
+          return;
+        }
+
+        // This is a struct.get that returns an itable type;
+        // Change to return the corresponding vtable type.
+        Builder builder(*getModule());
+        replaceCurrent(builder.makeStructGet(
+          0,
+          curr->ref,
+          parent.structInfoByITableType[curr->type.getHeapType()]
+            ->javaClass.getStruct()
+            .fields[0]
+            .type));
+      }
+    };
+
+    Rerouter rerouter(*this);
+    rerouter.run(getPassRunner(), &wasm);
+    rerouter.runOnModuleCode(getPassRunner(), &wasm);
+  }
+
+  // Modify the struct definitions adding the itable fields to the vtable and
+  // preserving the vtable field names.
+  void updateTypes(Module& wasm) {
+    class TypeRewriter : public GlobalTypeRewriter {
+      J2CLItableMerging& parent;
+
+    public:
+      TypeRewriter(Module& wasm, J2CLItableMerging& parent)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+
+      void modifyStruct(HeapType oldStructType, Struct& struct_) override {
+        if (parent.structInfoByVtableType.count(oldStructType)) {
+          auto& newFields = struct_.fields;
+
+          auto structInfo = parent.structInfoByVtableType[oldStructType];
+          // Add the itable fields to the beginning of the vtable.
+          auto it = structInfo->itable.getStruct().fields.rbegin();
+          while (it != structInfo->itable.getStruct().fields.rend()) {
+            newFields.insert(newFields.begin(), *it++);
+            newFields[0].type = getTempType(newFields[0].type);
+          }
+
+          // Update field names as well. The Type Rewriter cannot do this for
+          // us, as it does not know which old fields map to which new ones
+          // (it just keeps the names in sequence).
+          auto iter = wasm.typeNames.find(oldStructType);
```

Comment:
Should this always exist? ie. should we use at instead?

Reply:
Done. We assume we have names for the optimization at the beginning so it should be fine.